### ISSUE TITLE
Infer property type from its default value if not set explicitly

### ISSUE THREAD

**Issue by snigdha920** [mikro-orm/mikro-orm#4060]
**Describe the bug**

When we define a field with a default value, without explicitly specifying the type - either in TypeScript or in the `@Property` decorator, we get the following error while to create an entity:

Entity:

```
@Entity()
export class Author {
  @PrimaryKey()
  _id!: ObjectId;

  @SerializedPrimaryKey()
  public id!: string;

  @Property()
  name: string;

  @Property({ default: 0 })
  bestsellers = 0;

  constructor(name: string) {
    this.name = name;
  }
}
```

Error: 

```
"Trying to set Author.bestsellers of type 'string' to '0' of type 'number'"
```

**Stack trace**
```
"ValidationError: Trying to set Author.bestsellers of type 'string' to '0' of type 'number'",
            "    at Function.fromWrongPropertyType (/Users/snigdhasingh/nestjs-mikro-subscriptions-context/node_modules/.pnpm/@mikro-orm+core@5.6.11_@mikro-orm+mongodb@5.6.11/node_modules/@mikro-orm/core/errors.js:22:16)",
            "    at EntityValidator.validateProperty (/Users/snigdhasingh/nestjs-mikro-subscriptions-context/node_modules/.pnpm/@mikro-orm+core@5.6.11_@mikro-orm+mongodb@5.6.11/node_modules/@mikro-orm/core/entity/EntityValidator.js:65:44)",
            "    at /Users/snigdhasingh/nestjs-mikro-subscriptions-context/node_modules/.pnpm/@mikro-orm+core@5.6.11_@mikro-orm+mongodb@5.6.11/node_modules/@mikro-orm/core/entity/EntityValidator.js:24:35",
            "    at Array.forEach (<anonymous>)",
            "    at EntityValidator.validate (/Users/snigdhasingh/nestjs-mikro-subscriptions-context/node_modules/.pnpm/@mikro-orm+core@5.6.11_@mikro-orm+mongodb@5.6.11/node_modules/@mikro-orm/core/entity/EntityValidator.js:13:20)",
            "    at ChangeSetComputer.computeChangeSet (/Users/snigdhasingh/nestjs-mikro-subscriptions-context/node_modules/.pnpm/@mikro-orm+core@5.6.11_@mikro-orm+mongodb@5.6.11/node_modules/@mikro-orm/core/unit-of-work/ChangeSetComputer.js:38:28)",
            "    at UnitOfWork.findNewEntities (/Users/snigdhasingh/nestjs-mikro-subscriptions-context/node_modules/.pnpm/@mikro-orm+core@5.6.11_@mikro-orm+mongodb@5.6.11/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js:416:50)",
            "    at UnitOfWork.computeChangeSets (/Users/snigdhasingh/nestjs-mikro-subscriptions-context/node_modules/.pnpm/@mikro-orm+core@5.6.11_@mikro-orm+mongodb@5.6.11/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js:340:18)",
            "    at UnitOfWork.doCommit (/Users/snigdhasingh/nestjs-mikro-subscriptions-context/node_modules/.pnpm/@mikro-orm+core@5.6.11_@mikro-orm+mongodb@5.6.11/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js:254:18)",
            "    at processTicksAndRejections (node:internal/process/task_queues:95:5)"
```

**To Reproduce**
Reproduction repo with steps to reproduce: https://github.com/snigdha920/nestjs-mikro-validation

**Expected behavior**

I expect the type to be inferred from the TypeScript definition, or even the default value provided in the `@Property` decorator. 

**Additional context**

If I set `validate: false` in the options while initialising Mikro [here](https://github.com/snigdha920/nestjs-mikro-validation/blob/main/src/app.module.ts#L18), it just works.

This used to work as expected in 5.6.8, fails in 5.6.11.

**Versions**

| Dependency | Version |
| - | - |
| node | v19.6.1 |
| typescript | 4.7.4 |
| mikro-orm | 5.6.11 |
| your-driver | 5.6.11 |

**Comment by roderik** [comment]
FYI, a lot of the standard eslint typescript sets tell you you should not type it as it can be inferred, you can disable the error by doing `"@typescript-eslint/no-inferrable-types": "off"` but have not found a way to enforce it.

**Comment by snigdha920** [comment]
> But it's true we could use the default to infer the type. That is a feature request to me.

On more thought, I wouldn't add such a feature actually because it enables inconsistency in the code. In places we would have the `default` value set it would infer the type and in places where it is not set it would error out. It would be easier to just enforce type definition for class members everywhere.

> It should have never been working :] Sounds like a bug fix that uncovered wrong entity definition to me.

I see, and it makes sense because that is similar to how NestJS/GraphQL works for example.

**Comment by snigdha920** [comment]
> FYI, a lot of the standard eslint typescript sets tell you you should not type it as it can be inferred, you can disable the error by doing `"@typescript-eslint/no-inferrable-types": "off"` but have not found a way to enforce it.

Found it, it's this: https://typescript-eslint.io/rules/typedef#membervariabledeclaration

**Comment by snigdha920** [comment]
Perhaps we can add this ES Lint rule in the documentation: https://typescript-eslint.io/rules/typedef#membervariabledeclaration? Would be easier for the future.

**Comment by B4nan** [comment]
I actually do like the idea of inference from the default value. Especially if we use the property initializer, TS infers it the same way too.

**Comment by erie0210** [comment]
I'm interested. Can I take over this issue? @B4nan

**Comment by B4nan** [comment]
Sure

**Comment by erie0210** [comment]
Is it correct to do a default type inference for the mongo driver, or is there something else I should be aware of? @B4nan

**Comment by erie0210** [comment]
My plan is that in [`initDefaultValue` in `MetadataDiscovery`](https://github.com/mikro-orm/mikro-orm/blob/c047da00848de8da2060e29a1d7791526356201d/packages/core/src/metadata/MetadataDiscovery.ts#L945-L951), if there is a  `default`, infer the type value and put it in `this.type`.

**Comment by B4nan** [comment]
Closing as implemented via #4150
